[IV-25-26] Objetivo 4

gabrielherreraloz opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [ ] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

Closes #7
Closes #19 
Closes #20
Closes #22
Closes #24

gabrielherreraloz committed:
Creada base de Objetivo 4

gabrielherreraloz committed:
FIX-Readme sin palabras clave

gabrielherreraloz committed:
Creadas estructuras de datos(#22)

Se han creado distintas estructuras de datos para almacenar la información extraida

gabrielherreraloz committed:
Añadidos archivos HTML(#23)

Se han subido varios arcchivos HTML sobre los que realizar el scrappeo de la información

gabrielherreraloz committed:
FIX-Corregido iv.yaml (#22)

gabrielherreraloz committed:
FIX-Corregido iv.yaml (#24)

La ruta del archivo de tests era incorrecta

gabrielherreraloz committed:
Documentada eleccion de herramientas de test(#24)

gabrielherreraloz committed:
Modificados diccionarios con punteros(#22)

Se ha integrado el uso de punteros en los diccionarios para permitir su modificación desde otros archivos

Eliminado código erróneo

Se subió por error el archivo de tests

gabrielherreraloz committed:
Ampliada documentación de tests(#24)

gabrielherreraloz committed:
Extracción del número de línea(#19)

Se ha creado una función que toma como parametro el bloque del código html donde se encuentra la información y esta extrae solo el número de la línea separandolo del nombre completo de esta.

gabrielherreraloz committed:
Extracción de lso nombres de paradas(#20)

Se han creado dos funciones, la función principal toma como parámetro el bloque html donde se encuentra la información, extrae el trozo donde se especifican los nombres, y lo pasa a la segunda función creada, la cual exctrae cada nombre por separado, los almacena todos en un array de string y lo devuelve.

gabrielherreraloz committed:
Extracción de todos los horarios(#21)

Se han creado dos funciones, la primera extrae de un bloque html pasado como parámetroo, donde se encuentra la información, cada línea de la tabla, donde salen los horarios asociadas a cada parada, cada columna de horarios es una parada, por lo que devuelve una matriz de todos los horarios ordenadorrs por hora y parada.
La segunda función, recopila toda la información extraida y lo almacena en el struct Linea, donde almacena el nombre de la línea y sus horarios y paradas correspondientes.

gabrielherreraloz committed:
Creado archivo de tests(#19)

Se han creado distintas verificaciones de tests sobre el proyecto, tanto de casos típicos(HappyPath) como de casos raros(SadPath), en estos se testean casos como el funcionamiento con un html estandar de la web, con extractos de html vacíos, datos incompletos, etc.

gabrielherreraloz committed:
Modificado scrapper por tests(#19)

Se ha modificado el códgio de scrpaeo del HTML, ya que al probar los tests se ha identificado un potencial mal funcionamiento del proyecto a la hora de extraer el número de línea, pues se salía del tamaño admitido del vector cuando no había ningún titulo válido en el HTML.

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,48 @@
+
+### Criterios de selección
+   - Que no sea necesaria su instalación mediante dependencias externas.
+   - Velocidad de ejecución máxima.
```
¿Crees que es un criterio objetivo en tu proyecto? ¿Te está ayudando a determinar cuál es la mejor herramienta?

gabrielherreraloz replied:
La velocidad de ejecución sí me parece un criterio bastante objetivo.

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,48 @@
+
+### Criterios de selección
+   - Que no sea necesario instalarlo con dependencias externas, sería ideal si viene integrado junto con la herramienta de testing seleccionada.
+   - Sintaxis más simplificada, aunque predomina el criterio anterior por su importancia.
```
¿Es realmente un buen criterio si no se usa para decidir tu elección? ¿Estas realmente comparando las opciones con este criterio? ¿No crees que sea ruido?

gabrielherreraloz replied:
No creo que sea ruido ya que el primer criterio tiene más importancia, pero en caso de empate sobre el primero, els egundo podría decantar la balanza

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,48 @@
+La elección definitiva es la bibilioteca estandar de Go, ya que ofrece la herramienta directamente integrada en el lenguaje sin necesidad de instalaciones adicionales, y sacrificamos la simplicidad de la sintaxis ya que no es un factor tan determinante en el producto final como lo es el criterio anteriormente mencionado.
+
+## Herramienta CLI de ejecución de tests
+Se usará la propia del lenguaje ya que es la forma más optima de realizar esta tarea.
```
Entiendo que ese sea tu criterio, pero no crees que habría que dejar constancia de alternativas para dar a entender que has hecho un buen estudio del panorama de testing de go.

gabrielherreraloz replied:
Entiendo tu punto, pero siguiendo al pie de la letra la documentación de este objetivo, al estar directamente integrado en Go, no es necesarios barajar otras opciones, solo especificar que se ha elegido por eso.

gosema commented on pkg/scrapper/scrapper.go:
```diff
@@ -0,0 +1,130 @@
+    match := re.FindStringSubmatch(htmlContent)
+
+    if len(match) < 2 {
+        return ""
```
¿Es buena idea devolver una cadena vacía si no obtienes el número de línea? ¿Por qué no un error?

gosema commented on pkg/scrapper/scrapper.go:
```diff
@@ -0,0 +1,130 @@
+	idxVuelta := strings.Index(htmlContent, "<strong>Vuelta</strong>")
+	idxFinVuelta := strings.Index(htmlContent, `<div class="leyendas">`)
+
+	if idxIda == -1 || idxVuelta == -1 || idxFinVuelta == -1 {
```
De esta manera no puedes comprobar dónde da error exactamente, puede ser que el scrapeo falle porque no estan bien las idas, o las vueltas o las leyendas. Si tuvieses que arreglarlo no sabes especificamente cuál de los tres es el que falla

gosema commented:
Buenos tardes @gabrielherreraloz entiendo que quieras implementar la lógica de negocio de tu proyecto pero creo que (basándonos en el desarrollo ágil y siempre buscar el producto minimamente viable) estás tratando de resolver un problema demasiado completo y relacionado con la solución final.
Es cierto que necesitamos lo horarios, las paradas y las líneas ¿pero todo esto es tu primera fase sobre la que construir la lógica de negocio? Es decir, si un cliente quisiese tener un producto con valor del proyecto en esta fase de desarrollo no preferiría una versión más completa y correcta de como extraer las paradas o quizás que líneas hay antes que todo la lógica de scrapeo completa. 
No quiero decir que tu objetivo sea incorrecto, sino que de esta forma facilitaría seguir las buenas prácticas y el desarrollo poco a poco de código de calidad. Dale una vuelta y dime lo que opinas.

gabrielherreraloz committed:
FIX-Código reducido a PMV(#19)

Se han eliminado funciones del código por superar los límites de un PMV y se dejan dichas funcionalidades para el futuro. Se ha eliminado el código que extraía todos los horarios de las líneas y ahora unicamente extrae el nombre y las paradas de la línea.

gabrielherreraloz committed:
FIX-Implementados mensajes de error(#19)

Se ha implementado la devolución de errores por parte de las funciones de scrapping para ayudar a la identificación de fallos en este, siguiendo los comentarios de @gosema.

gabrielherreraloz committed:
FIX- Ajustados Tests a nuevo código(#20)

COn la reducción del código de scrapping, se han eliminado los tests sobre la matriz de horarios y se han corregido los mensajes de error al objtener el número de línea.

gabrielherreraloz committed:
Actualizada documentación de tests(#24)

Siguiendo los comentarios de @gosema, se han suprimido criterios de selección irrelevantes, además, se han ordenado las herramientras de menor a mayor abstracción.

gabrielherreraloz committed:
Modificados tests usando htmls(#19)

Se han modificado los tests para realizar las comprobaciones de Happy Path usando directamente los HTMLs descargados en local, se han agrupado todos los tests de happy path en un solo bloque y a continuación, todos los tests de sad path.

gabrielherreraloz commented:
@gosema cambios realizados.

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,49 @@
+Se han tenido en cuenta las dos herramientas más usadas en proyectos con Go:
+
+   **Estandar de Go**: 
+   Instalada nativamente junto con el lenguaje.
```
¿Qué opinas sobre que Go no tenga biblioteca de aserciones? En vez de funciones como expect() o assert() promueve comparaciones con el uso de if y el uso de t.error.
A lo mejor sería interesante destacar esta información en la documentación de alguna manera.

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,49 @@
+
+
+## Herramienta CLI de ejecución de tests
+Se usará la propia del lenguaje ya que es la forma más optima de realizar esta tarea.
```
Quizá hablar algo sobre que go test es tanto la CLI como el testrunner

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,49 @@
+   Muy parecido a GoTest, pero muestra la información más legible, sin embargo requiere la instalación de alguna herramienta adicional. No escribe en el directorio del código fuente.
+
+   **ginkgo**: https://pkg.go.dev/github.com/onsi/ginkgo/ginkgo
+   No está instalado de forma nativa junto con el lenguaje, por lo que inclumple el principal criterio además de tener riesgo de presentar casos en los que sí escriba en el directorio de código fuente.
```
Quizá convendría detallar un poco más que "casos" son en los que se escribe

gosema commented on pkg/scrapper/scrapper.go:
```diff
@@ -0,0 +1,100 @@
+func ExtraerNombresParadas(bloqueHTML string) []string {
+    var nombres []string
+
+    re := regexp.MustCompile(`(?s)<th[^>]*>(.*?)</th>`)
```
Te comento lo mismo que observó el profesor en mi código, las expresiones regulares al igual que el código deben ser claras. Ya se por el uso de comentarios, commits bien detallados o herramientas de espacios en blanco no significativos para desglosar el significado de la expresión.
Quizá analizando esto encuentres algo que sobra de la expresión o que podría hacerse de una manera más clara

gosema commented on pkg/utils/utils_test.go:
```diff
@@ -0,0 +1,159 @@
+        fileName := filepath.Base(filePath) 
+
+        t.Run("Happy_Path_File_" + fileName, func(t *testing.T) {
+            t.Parallel() 
```
¿Estás haciendo buen uso del parallel en el bucle leyendo así los ficheros?

gosema reviewed (changes requested):
Hay algunos commits que realizas asociados al issue #25 sobre todo en el archivo referido a  tests. Como dijo el profesor en clase los test deben salir de un buen issue del problema.
De la forma en la que lo haces ¿estarías justificando porque has creado cada test?

gabrielherreraloz committed:
Mejorada documentaciónde Tests(#24)

Siguiendo de nuevo los comentarios de @gosema, se ha mejorado la especificación de la documentación, recalcando el uso de sintaxis que hace GoTest, apuntando mejor que esta herramienta actúa como Test Runner y Herramienta CLI, y especificando los casos en los que GinkGo podría actuar de forma contraria a la deseada.

gabrielherreraloz committed:
FIX- Mejora documentación(#24)

No se había actualizado correctamente toda la documentación.

gabrielherreraloz committed:
Añadidos comentarios sobre scrapeo(#20)

Se ha detallado mejor la explicación sobre el código de scrapeo con el uso de comentarios en las partes más críticas y de búsqueda de expresiones regulares para facilitar su revisión.

gabrielherreraloz committed:
Eliminado paralelismo en tests(#19)

Se ha eliminado el paralelismo en la ejecución de tests ya que tras hacer pruebas, el tiempo de ejecuicón es el mismo con o sin este, y no implementarlo fomenta la simplicidad del código.

gosema commented:
> Hay algunos commits que realizas asociados al issue #25 sobre todo en el archivo referido a tests. Como dijo el profesor en clase los test deben salir de un buen issue del problema. De la forma en la que lo haces ¿estarías justificando porque has creado cada test?

Siguen sin haberse llevado las buenas prácticas de issues de los que nacen los tests

gabrielherreraloz commented:
@gosema hecho

gosema commented:
Por mi parte lo veo bien, tanto la justificación de los tests, los issues creados, la implementación de la lógica de negocio y los tests realizados.

gabrielherreraloz commented:
@jj listo para revisión

JJ commented on docs/Tests.md:
```diff
@@ -0,0 +1,51 @@
+   **Go Test**: https://pkg.go.dev/testing
+   La herramienta nativa de Go para testing, está instalado nativamente junto con el propio lenguaje, por lo que no es necesaria su instalación aparte. No escribe en el directorio del código fuente.
+
+   **gotestsum**: https://pkg.go.dev/github.com/IstrateM/gotestsum/pkg/gotestsum 
```
No es un test runner. Por favor mira la descripción.

JJ commented on docs/Tests.md:
```diff
@@ -0,0 +1,51 @@
+Se han tenido en cuenta las dos herramientas más usadas en proyectos con Go:
+
+   **Estandar de Go**: 
+   Instalada de forma nativa junto con el lenguaje.
```
¿Qué biblioteca de aserciones usa Go, en concreto?

JJ commented on docs/Tests.md:
```diff
@@ -0,0 +1,51 @@
+   Requiere de su instalación adicional ya que no está integrado en Go.
+
+### Elección final
+La elección definitiva es la bibilioteca estandar de Go, ya que ofrece la herramienta directamente integrada en el lenguaje sin necesidad de instalaciones adicionales. 
```
¿Qué herramienta usa Go?

JJ commented on docs/Tests.md:
```diff
@@ -0,0 +1,51 @@
+### Elección final
+La elección definitiva es la bibilioteca estandar de Go, ya que ofrece la herramienta directamente integrada en el lenguaje sin necesidad de instalaciones adicionales. 
+
+Cabe destacar, solo como apunte extra sin entrar en los criterior tomados, que mientras muchas bibliotecas hacen uso de assert(), Go promueve el uso de if() y t.Errorf para reportar errores en los tests, lo cual permite el lanzamiento de errores más personalizados y descriptivos para cada ocasión.
```
Entonces, ¿qué herramienta usaría?

JJ reviewed (commented):
Aparte de lo dicho:
- ¿Alguna de las herramientas que mencionas tiene también un CLI para ejecutarla?
- ¿Cuál es la única fuente de verdad para los elementos del dominio que has usado?

gabrielherreraloz committed:
Ampliados detalles de documentación(#24)

Se ha especificado mejor qué bilioteca de aserciones utiliza Go nativamente, además se ha eliminado gotestsum como opción de TestRunner por no cumplir la descripción exactamente de lo que consideramos como TestRunner.

gabrielherreraloz committed:
Especificadas herramientras CLI(#24)

Se ha especificado que hay más herramientas que vienen con herramienta CLI propia incorporada.

gabrielherreraloz commented:
> Aparte de lo dicho:
> 
> * ¿Alguna de las herramientas que mencionas tiene también un CLI para ejecutarla?
> * ¿Cuál es la única fuente de verdad para los elementos del dominio que has usado?

Para la única fuente de verdad, está todo almacenado en dos map, que están dentro del struct Almacen.
He seguido la misma estructura de la que fue aprobada del objetivo 2 de chelunike. No sé si es a eso a lo que te refieres.

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,49 @@
+Se han tenido en cuenta las dos herramientas más usadas en proyectos con Go:
+
+   **Estandar de Go** (testing): https://pkg.go.dev/testing
+   Instalada de forma nativa junto con el lenguaje, el cual utiliza "testing" comp paquete
```
Quizá lo que quiere decir el profesor cuando te ha comentado esta línea es que Go no incluye una bibliote de aserciones en su estándar

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,49 @@
+### Elección final
+La elección definitiva es la bibilioteca estandar de Go, testing, ya que ofrece la herramienta directamente integrada en el lenguaje sin necesidad de instalaciones adicionales. 
+
+Cabe destacar, solo como apunte extra sin entrar en los criterior tomados, que mientras muchas bibliotecas hacen uso de assert(), testing promueve el uso de if() y t.Errorf para reportar errores en los tests, lo cual permite el lanzamiento de errores más personalizados y descriptivos para cada ocasión.
```
Entiendo que promueva el uso de if, pero "t.Errorf" es del paquete testing. El cual no considero que sea realmente una biblioteca de aserciones

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,49 @@
+
+
+## Herramienta CLI de ejecución de tests
+Se usará la propia del lenguaje ya que es la forma más optima de realizar esta tarea. Se ha seleccionado Go Test no solo como herramienta CLI, sino como Test Runner, como se ha indicado anteriormente, ejecutando esta herramienta ambas funciones, manteniendo la consistencia durante todo el test.
```
¿Óptima en base a qué?

gosema replied:
No entiendo esto, entonces go test que es? una biblioteca de aserciones? test runner? CLI?

gosema commented on docs/Tests.md:
```diff
@@ -0,0 +1,49 @@
+
+## Herramienta CLI de ejecución de tests
+Se usará la propia del lenguaje ya que es la forma más optima de realizar esta tarea. Se ha seleccionado Go Test no solo como herramienta CLI, sino como Test Runner, como se ha indicado anteriormente, ejecutando esta herramienta ambas funciones, manteniendo la consistencia durante todo el test.
+En caso de haber elegido ginkgo, podríamos heber utilizado su herramienta CLI propia pues este también la incluye.
```
No crees que estás sesgando tu opinión, me da la impresión que descartas ginkgo porque ya lo habías hecho antes como test runner. Pero a lo mejor habría considerar si como herramienta CLI puede aportar algunas ventajas según tus criterios (no hay) para recuperarlo del test runner.

gabrielherreraloz replied:
No, solo lo menciono como extra, claramente la opcion de usar Go es la mejor porque es la integrada del lenguaje. Solo he seguido las indicaciones de la documentación, cito textualmente:

"hay lenguajes o marcos de programación, como Go o deno, que ya incluyen una suborden test que realiza esta labor; en esto casos no habrá que establecer, como es natural, un criterio para elegirlo, porque la mejor práctica es usar la que ya hay, pero si habrá que dejar claro en la documentación del objetivo que este es el caso y que se va a hacer de esa forma."

He explicado que lo elijo por esa razón, como dice en el guión.

JJ replied:
Hay más bibliotecas de asrciones en Go, por cierto. De veras que no os cuesta buscar un poco usando el viejo Google o las herramientas habituales, StackOverflow o Reddit.

gabrielherreraloz replied:
> Hay más bibliotecas de asrciones en Go, por cierto. De veras que no os cuesta buscar un poco usando el viejo Google o las herramientas habituales, StackOverflow o Reddit.

De hecho, sí que las he buscado en google, y en reddit, donde lo que más se comentaba era testify, que es la más usada en Go según tengo entendido, y testing, que no es una biblioteca de aserciones como tal, pero permite hacer lo que una biblioteca de aserciones propia haría. 

Todo esto lo digo desde el desconocimiento que pueda tener yo, claro, o que haya podido entender mal el concepto.

gabrielherreraloz commented:
> > Aparte de lo dicho:
> > 
> > * ¿Alguna de las herramientas que mencionas tiene también un CLI para ejecutarla?
> > * ¿Cuál es la única fuente de verdad para los elementos del dominio que has usado?
> 
> Para la única fuente de verdad, está todo almacenado en dos map, que están dentro del struct Almacen. He seguido la misma estructura de la que fue aprobada del objetivo 2 de chelunike. No sé si es a eso a lo que te refieres.

@JJ si me pudieras aclarar esto, por favor

JJ commented on pkg/scrapper/scrapper.go:
```diff
@@ -0,0 +1,122 @@
+package scrapper
```
¿scrapper se escribe así?

JJ commented on pkg/scrapper/scrapper.go:
```diff
@@ -0,0 +1,122 @@
+    match := re.FindStringSubmatch(htmlContent)
+
+    if len(match) < 2 {
+        return "", fmt.Errorf("el contenido HTML no contiene el encabezado <h2> necesario para el título")
```
No has modelizado los errores. Mira [esto](https://stackoverflow.com/questions/56580030/can-we-create-subtypes-for-errors-in-go) para una posible solución. [Aquí](https://pkg.go.dev/errors) habla de un paquete que lo hace y finalmente en [estos ejemplos](https://gobyexample.com/errors) explica cómo se definen errores específicos en Go.

JJ reviewed (commented):
Comentarios adicionales.

JJ commented:
> > > Aparte de lo dicho:
> > > 
> > > * ¿Alguna de las herramientas que mencionas tiene también un CLI para ejecutarla?
> > > * ¿Cuál es la única fuente de verdad para los elementos del dominio que has usado?
> > 
> > 
> > Para la única fuente de verdad, está todo almacenado en dos map, que están dentro del struct Almacen. He seguido la misma estructura de la que fue aprobada del objetivo 2 de chelunike. No sé si es a eso a lo que te refieres.
> 
> @JJ si me pudieras aclarar esto, por favor

¿Dónde estás usando esa fuente de verdad?

gabrielherreraloz committed:
FIX-Corregido nombres scraper

Se estaba nombrando mal la palabra scraper todo el rato, llamandola scrapper, se ha corregido.

JJ commented:
Más sobre [errores en Go](https://www.reddit.com/r/golang/comments/y80j1f/error_subtypes/)

gabrielherreraloz committed:
Mejorada documentación Tests(#24)

Se ha mejorada la explicación y especificación de qué es Go Test y el paquete testing exactamente, además, se han añadido mas opciones como biblioteca de aserciones.

JJ commented:
Buscando go assert te da al menos 2 bibliotecas más en la primera página.

gabrielherreraloz commented:
> Buscando go assert te da al menos 2 bibliotecas más en la primera página.

yo he buscado "biblioteca de aserciones golang".
Lo he ido sacando de aquí: https://www.reddit.com/r/golang/comments/1h1v0eo/exploring_all_assertion_libraries_test_runners/?tl=es-es

gabrielherreraloz committed:
Delete pkg/scrapper/scrapper.go

gabrielherreraloz committed:
Realizada modelización de errores(#19)

Se han modificado las funciones de scraping, para que, siguiendo la estructura típica de Go, devuelva siempre el resultado y un error. Estos errores posibles se han definido en un nuevo archivo destinado a definirlos.

gabrielherreraloz committed:
Creados nuevos errores(#19)

Se han creado e implementado nuevos tipos de error en el código de scrapeo

gabrielherreraloz committed:
Modificada lógica de tests(#20)

Se ha implementado la lógica de comprobación de errores correctos del scrapeo.

gabrielherreraloz committed:
Creados tests para única fuente de verdad(#19)

Se han creado tests para segurar que no se guardan líneas repetidas

gabrielherreraloz committed:
Se ha modificado el codigo para pasar los tests(#20)

gabrielherreraloz commented:
@gosema revisión

gosema commented:
@gabrielherreraloz Lo veo bien

gabrielherreraloz commented:
@JJ listo para revisión

JJ reviewed (approved):
👍 Tendrías qu emeterle un poco de "Don't repeat yourself", porque has repetido un montón de código, pero ya a estas alturas...
